File: 2021/src/day01/day01.swift
import Foundation

enum Day1 {
  static func parse(_ contents: String) -> [Int] {
    return contents.split(separator: "\n").map { Int($0) ?? -1 }
  }

  static func part1(input: String) -> Int {
    let data = parse(input)
    return data.indices
      .filter { $0 > 0 && data[$0] > data[$0 - 1] }
      .count
  }

  static func part2(input: String) -> Int {
    let data = parse(input)
    let windows = data.indices
      .filter { $0 < data.count - 2 }
      .map { data[$0] + data[$0 + 1] + data[$0 + 2] }

    return windows.indices
      .filter { $0 > 0 && windows[$0] > windows[$0 - 1] }
      .count
  }
}

File: 2021/src/day01/day01.js
const parseInput = (rawInput) => rawInput.trim().split("\n").map(Number);

export const part1 = (rawInput) => {
  const input = parseInput(rawInput);
  return input.filter((v, i) => (i === 0 ? false : v > input[i - 1])).length;
};

export const part2 = (rawInput) => {
  const input = parseInput(rawInput);

  return parseInput(rawInput)
    .slice(0, input.length - 2)
    .map((_, i) => input.slice(i, i + 3).reduce((sum, v) => sum + v, 0))
    .filter((v, i, all) => (i === 0 ? false : v > all[i - 1])).length;
};
